### PR TITLE
commit to mods-batch#3

### PR DESCRIPTION
**Pull request by oormicreations**
### 31-Jan-2021

Update readme
_File : README.md_

Better placement of "Read More" button
_File: routes/Home/index.tsx Line:333_

Button to set default time span
_File: routes/Edit/VideoSplitter.tsx Line: 541_

Set start time to 0 if endtime for previous segment == duration
_File: routes/Edit/VideoSplitter.tsx Line: 148_

### 01-Feb-2021

Experimental tags
_File: src/components/Tags.tsx, TagsSuggestions.tsx, VideoSplitter.tsx Line: 661_

```
//Dependencies for drag n drop tags with suggestions
//npm install --save react-tag-input
//npm install --save react-dnd@5.0.0
//npm install --save react-dnd-html5-backend@3.0.2
```

### 03-Feb-2021

Deleted copies of images. All images are in one folder.
_File: src/images_

### CODE REVIEW

File: src/components/TagSuggestions.tsx
Comment (JKHeadley) [inline src/components/TagSuggestions.tsx:1]: Wow, nice list! haha

Just curious, where did this list come from? Is it from an index of the MBT website or book?

If we decide to implement autocomplete, we will want to make sure to include a list of current tags from the database as well.  There are definitely pros and cons to this feature that we will need to discuss.

**Comment by oormicreations** [comment]
Updates : 06/02/2021

Button to set end time of segment where video is paused.

Use uuid to set tag index

Capitalise first letter of the tag.

New tags with suggestions
_File: src/components/Tags.tsx, TagsSuggestions.tsx, VideoSplitter.tsx_ 

Fix: Backspace deleting tags.

Tip component
_File: src/components/Tip.tsx_

File: src/components/TagSuggestions.tsx
Comment (oormicreations) [reply src/components/TagSuggestions.tsx:1]: Thanks :smile: 
I typed it out. I am typing these words since many months everyday !
We can add a few more from the data surely.

**Comment by oormicreations** [comment]
**07-Feb-2021**

- Clear title field, no need to select and delete it.
- Fix video title cutting off. Reduced sidebar width.
- Search thumbnails are now small.
- Changed logo to square.
- Validation while saving segment
- Tried a contact page. ToDo : Do conditional routing (or block default page?)

**Comment by JKHeadley** [comment]
@oormicreations thanks for all the updates! I still have not been able to review everything but I will keep at it.

For now, lets not add any more changes to this PR. In general, it would be good to keep each PR to a small manageable set of changes. Usually each PR will refer to a single issue. This will keep the PR's easier to review and help us merge each individual change in sooner. Let me know if you have any thoughts or questions on this. Thanks so much!

**Comment by oormicreations** [comment]
Sure, I will do a new PR for each set of changes.

**Comment by oormicreations** [comment]
@JKHeadley 
I have some more changes ready to push. I can push it in a new branch and do a PR for that, but like before there will be duplicate changes (same as in this PR).
What should I do?

File: src/App.css
Comment (JKHeadley) [inline src/App.css:38]: This looks much better, thanks!

File: src/components/LandingPage.tsx
Comment (JKHeadley) [inline src/components/LandingPage.tsx]: It will be most convenient to use the `hasPermission` utility to check for volunteers. See here for an example:

https://github.com/cusac/mbt.guide/blob/e81cef8e003e31e0b259652da9ca12457e0213d4/src/components/SegmentViewer.tsx#L24

By default we are not including the user.groups object since it does not contain information that we need for now. The `Volunteer` string will be included in the user's scope as long as they are a part of that group.

You can see these docs for more insight into user scopes if you like:

https://github.com/JKHeadley/appy-backend/wiki/Authorization

Note that the permission system we are using is based off of that repository.

File: src/components/Tags.tsx
Comment (JKHeadley) [inline src/components/Tags.tsx]: Note that this uses the class style of React components whereas most of the components in this app use the function style. Both are fine, however I have found that the function style components are much more flexible and lightweight and easier to manage.

File: src/components/Tags.tsx
Comment (JKHeadley) [inline src/components/Tags.tsx]: I apologize if I have not mentioned this earlier, but we will soon be enforcing that all tags are all lower case.  This is part of an effort to normalize the tags and reduce/combine duplicate tags. I'm sure you have already noticed that both `Fear` and `fear` are individually in the top 10 most used tags. These need to reference the same tag, so we have to create a normalization scheme, and lowercasing all tags is the simplest/most bug free approach.

Frank K. is currently working on scripts to enforce this and to also retroactively enforce this rule on all existing tags.

File: src/components/Tags.tsx
Comment (JKHeadley) [inline src/components/Tags.tsx]: No, these are handled by the backend and should not be included.

File: src/components/Tags.tsx
Comment (JKHeadley) [inline src/components/Tags.tsx]: Overall this is a very cool component. However I don't believe we should attempt to merge this to the production site for two main reasons:

1) Tags cannot be drag and dropped between ranks, and rearranging tags within a rank does not serve a functional purpose.
2) Duplicate tags are not detected, which can be confusing.

I think this is a great start at improving the tag experience. My suggestion is that we move this feature to a separate branch and remove it from this PR so that it does not prevent us from merging the other features.

File: src/components/Tip.tsx
Comment (JKHeadley) [inline src/components/Tip.tsx:19]: Nice and handy component!

File: src/components/VideoItem.tsx
Comment (JKHeadley) [inline src/components/VideoItem.tsx:12]: Did not know about Fragments! Very useful component, thanks!

File: src/components/YouTubePlayer.tsx
Comment (JKHeadley) [inline src/components/YouTubePlayer.tsx:116]: Could you explain the purpose of this line?

File: src/routes/Contact/index.tsx
Comment (JKHeadley) [inline src/routes/Contact/index.tsx]: This is great! I'm not familiar with how this form works, would you mind linking some documentation for me or point me to where the contact reports end up?

File: src/components/YouTubePlayer.tsx
Comment (JKHeadley) [reply src/components/YouTubePlayer.tsx:116]: Ah, I see that this is related to the "Set End Time" button. Very handy! I will have to study the code more but I suspect that we can use our state management functions to handle this rather than the global variable, however this seems to work for now.

File: src/routes/Edit/VideoSplitter.tsx
Comment (JKHeadley) [inline src/routes/Edit/VideoSplitter.tsx]: If possible lets try to clean up/remove any commented log lines.

File: src/routes/Edit/VideoSplitter.tsx
Comment (JKHeadley) [inline src/routes/Edit/VideoSplitter.tsx]: This is nice and definitely needed. If possible it would be nice to integrate a more formal form validation process, however this is perfectly fine for now.

File: src/routes/Edit/VideoSplitter.tsx
Comment (JKHeadley) [inline src/routes/Edit/VideoSplitter.tsx]: Unless you believe logs will be helpful in production lets try not to merge them into master.

File: src/routes/Edit/VideoSplitter.tsx
Comment (JKHeadley) [inline src/routes/Edit/VideoSplitter.tsx]: Lets go ahead and revert to the old tag system for now and move these features to a new branch until this functionality is more mature.

File: src/routes/Edit/VideoSplitter.tsx
Comment (JKHeadley) [inline src/routes/Edit/VideoSplitter.tsx]: Haha, looks like we are on the same page. Since we are not merging this feature we will need to remove if from this branch/PR and move it to it's own feature branch.

File: src/routes/Home/index.tsx
Comment (JKHeadley) [inline src/routes/Home/index.tsx]: For future reference, passing properties into components is fine, however if the property is already managed by Redux (such as the current user) it might be cleaner to retrieve the data from Redux within the component rather than passing it as a property. For example, see:

https://github.com/cusac/mbt.guide/blob/e81cef8e003e31e0b259652da9ca12457e0213d4/src/components/SegmentViewer.tsx#L11

If you like at some point we can do an overview of Redux (state management) within this app. The basics would not take long to cover.

**Review by JKHeadley — CHANGES_REQUESTED** [review]
Overall great changes! Please review the comments and requested changes and respond as needed (and resolve any conversations that you are satisfied are complete).

**Comment by JKHeadley** [comment]
> @JKHeadley
> I have some more changes ready to push. I can push it in a new branch and do a PR for that, but like before there will be duplicate changes (same as in this PR).
> What should I do?

Thanks for asking. Yes this is a common situation that can seem tricky to handle appropriately.

Often times you will submit a PR, and continue development on a new branch soon after. Ideally this new branch will be independent of the changes made in the previous PR and so the branch will be a new branch fresh off of the master branch. This keeps development simple and easy to manage, and is why it is best to work on a single feature in a branch.

However, sometimes (as in this case) the next feature you work on will depend on the changes that are currently being reviewed in the PR. In this case, I see that you have two options:

1) Wait for the dependent PR to be merged into master, so that you can cleanly start a new branch off of master with the required new features.

2) Branch off of the branch being reviewed to continue development. In this case it will be important that you continually merge in any changes made to the branch being reviewed to stay in sync, and once that branch is merged into master you will need to merge master into your new branch.

I think option 2 can work fine (and it is usually the option I take since I am inpatient, haha), however it can sometimes result in somewhat unintuitive merging conflicts. This is another reason small branches/PRs are optimal since they reduce the confusion when merging conflicts occur.

In general, it is a good practice to continually merge any updates from master into your current branch to reduce conflicts and keep in sync. Finally, merging master into a PR branch is always the last step before merging into master to guarantee that no conflicts are commited to master/production.

I hope this helps, feel free to reach out if this is not clear. We can always discuss this further in a chat.

**Comment by JKHeadley** [comment]
One final note so that it is clear. It is NOT recommended to copy/paste features from one branch to another to carry over functionality as this is almost guaranteed to cause conflicts. When functionality is needed from an existing branch, it is much better to create a new branch based off of that branch rather than creating a branch off of master and copying code over.

File: src/components/Tags.tsx
Comment (oormicreations) [reply src/components/Tags.tsx]: Sure, it can be done. We can test it out as it is before I write a new component.

File: src/components/Tip.tsx
Comment (oormicreations) [reply src/components/Tip.tsx:19]: Thank you. Will be useful in the forked version too. Intention is to get rid of instruction manuals.

File: src/components/VideoItem.tsx
Comment (oormicreations) [reply src/components/VideoItem.tsx:12]: Ya great thing to pile up stuff.

File: src/components/YouTubePlayer.tsx
Comment (oormicreations) [reply src/components/YouTubePlayer.tsx:116]: Believe me this cut down the time required to do stamping to half. Now mostly there is no need to set time manually.
The drop down tags have reduced the time, no need to type anything except title and description.
I added a new issue about the global var, so we don't forget.

File: src/routes/Contact/index.tsx
Comment (oormicreations) [reply src/routes/Contact/index.tsx]: The responses are emailed to the owner, and are also saved in a google sheet here:
https://docs.google.com/spreadsheets/d/1xVLwN8bPslgq-JtmKJKWd_JeIHglklfUeNscE-Bb1YA/edit#gid=1364178731
We can change this to any form, any email and any sheet if needed (We just change the form url). 
I have added your email (so you can test it).
I was thinking of coding it out, but we need backend to store the data, before we do that.

File: src/routes/Edit/VideoSplitter.tsx
Comment (oormicreations) [reply src/routes/Edit/VideoSplitter.tsx]: Done...

File: src/routes/Edit/VideoSplitter.tsx
Comment (oormicreations) [reply src/routes/Edit/VideoSplitter.tsx]: We will do that. Issue created.

File: src/routes/Edit/VideoSplitter.tsx
Comment (oormicreations) [reply src/routes/Edit/VideoSplitter.tsx]: Sure. I removed many. Thanks.

File: src/routes/Home/index.tsx
Comment (oormicreations) [reply src/routes/Home/index.tsx]: It is done. Thanks a lot for the hint. Yes, I need to learn more.

File: src/components/LandingPage.tsx
Comment (oormicreations) [reply src/components/LandingPage.tsx]: It is done. Thanks a lot. Its simpler. We have separate pages now.

File: src/components/Tags.tsx
Comment (oormicreations) [reply src/components/Tags.tsx]: It is done. We force lowercase. 
I don't know what happens to tags like LCS or AUM ;-)

File: src/components/Tags.tsx
Comment (oormicreations) [reply src/components/Tags.tsx]: I made drag and drop as a separate feature. Removed the in-rank dragging.
Detecting duplicates now.
Hopefully you can reconsider merging. If not, we work more !

File: src/routes/Edit/VideoSplitter.tsx
Comment (oormicreations) [reply src/routes/Edit/VideoSplitter.tsx]: Yes, I wanted you to see it first before merging. But now it is no more experimental.

**Comment by oormicreations** [comment]
> > @JKHeadley
> > I have some more changes ready to push. I can push it in a new branch and do a PR for that, but like before there will be duplicate changes (same as in this PR).
> > What should I do?
> 
> Thanks for asking. Yes this is a common situation that can seem tricky to handle appropriately.
> 
> Often times you will submit a PR, and continue development on a new branch soon after. Ideally this new branch will be independent of the changes made in the previous PR and so the branch will be a new branch fresh off of the master branch. This keeps development simple and easy to manage, and is why it is best to work on a single feature in a branch.
> 
> However, sometimes (as in this case) the next feature you work on will depend on the changes that are currently being reviewed in the PR. In this case, I see that you have two options:
> 
> 1. Wait for the dependent PR to be merged into master, so that you can cleanly start a new branch off of master with the required new features.
> 2. Branch off of the branch being reviewed to continue development. In this case it will be important that you continually merge in any changes made to the branch being reviewed to stay in sync, and once that branch is merged into master you will need to merge master into your new branch.
> 
> I think option 2 can work fine (and it is usually the option I take since I am inpatient, haha), however it can sometimes result in somewhat unintuitive merging conflicts. This is another reason small branches/PRs are optimal since they reduce the confusion when merging conflicts occur.
> 
> In general, it is a good practice to continually merge any updates from master into your current branch to reduce conflicts and keep in sync. Finally, merging master into a PR branch is always the last step before merging into master to guarantee that no conflicts are commited to master/production.
> 
> I hope this helps, feel free to reach out if this is not clear. We can always discuss this further in a chat.

@JKHeadley Thanks for clearing it up. Ya it is really confusing.
So from now on I will keep a set of related changes in one branch. All based on master.

The good news is that I have finished messing with main source code, and the rest will be new components, separate files. Merging will be cleaner.
I have created many more issues, but nothing to worry, there is nothing serious there, the app is working fine. I noted them as I found them while working.
I am also slowing down on this project now, because at this time I am doing 6-7 projects at the same time ! You will get a lot of time to review between updates (if any at all).

File: src/components/Tags.tsx
Comment (oormicreations) [reply src/components/Tags.tsx]: Thanks. Removed them.

File: src/routes/Edit/VideoSplitter.tsx
Comment (oormicreations) [reply src/routes/Edit/VideoSplitter.tsx]: I have separated the drag and drop feature for now. We will implement it later. It is now working fine, and implements the drop down suggestion feature mainly.
I left the old code there, so if you stick to this decision, we will revert back quickly.

File: src/App.tsx
Comment (JKHeadley) [inline src/App.tsx]: The need for a "videoId" parameter is a result of bad route structuring from early app development. It looks like there are still some issues when I visit the contact page.

I'm working to resolve these route issues in my current branch, so I will try to finish that branch soon and merge in the changes here before we merge this branch into production.

File: src/components/LandingPage.tsx
Comment (JKHeadley) [inline src/components/LandingPage.tsx]: I'm not sure that the word `Scope` will have much meaning to users and could be confusing. I'm guessing that you are intending to display a group name that the user is in, however I don't think there is a guarantee that the index you are referring to will contain that name. Could you clarify what information you would like to display to the user here?

File: src/components/TagSuggestions.tsx
Comment (JKHeadley) [reply src/components/TagSuggestions.tsx:1]: If we are enforcing lowercase we will probably want to make sure this list is lowercase as well to avoid confusion.

Comment (JKHeadley) [reply src/components/TagSuggestions.tsx:1]: Also, I think it is important to mention, the concept of suggesting tags has been discussed several times in the past. I have so far decided not to implement this feature for one main reason. I am concerned that providing suggestions will introduce a bias toward tag names and prevent users from accessing their own intuition/creativity to create the tags they feel most describe the segment since it is much easier to select a tag from a list than to create a new one.

I definitely understand this could help volunteers such as yourself who do many videos, however I am cautious to provide this feature to all users. If we do provide this to volunteers, I think it would be a good idea to encourage them to make create new tags if they don't find a good fit from the suggestions.

File: src/App.tsx
Comment (oormicreations) [reply src/App.tsx]: ok. It still tries to get videos and shows an error.

File: src/components/LandingPage.tsx
Comment (oormicreations) [reply src/components/LandingPage.tsx]: I was trying to display their group.

File: src/components/TagSuggestions.tsx
Comment (oormicreations) [reply src/components/TagSuggestions.tsx:1]: So now that you have rewritten the component, it seems you have decided to keep the suggestions. Then we are ready to merge this PR.
Anyway, if you like we can have a toggle switch to disable the suggestions and keep it off by default.

File: src/App.tsx
Comment (oormicreations) [reply src/App.tsx]: Is it done in the last PR?

Comment (JKHeadley) [reply src/App.tsx]: No this isn't done yet, sorry. I will focus on getting the PR with this fix finished 😄

File: src/components/LandingPage.tsx
Comment (JKHeadley) [reply src/components/LandingPage.tsx]: Okay, for now maybe we can remove this and later we can display a list of groups the user belongs to in their profile page, or if you think it would be really helpful we can make another PR later that fetches the user's groups and displays it.

File: src/App.tsx
Comment (oormicreations) [reply src/App.tsx]: No problem. No hurry.

File: src/components/LandingPage.tsx
Comment (oormicreations) [reply src/components/LandingPage.tsx]: Removed. Edited the file here, the deploy preview looks ok.

**Comment by JKHeadley** [comment]
@oormicreations merged in the latest from master and cleared up any conflicts. Should be good to go! If you can do a last check to maker sure I didn't mess anything up then I'll let you do the honors of merging it to master!

This ended up being quite the PR, haha (a month long? 😲 ). Smaller PRs moving forward will definitely help them get merged sooner. Thanks for all the amazing work!

**Comment by oormicreations** [comment]
@JKHeadley Amazing !

I was trying to see the deploy preview.
But seems to be a problem here. Build failed.

> /opt/build/repo/src/App.tsx
> 5:59:15 AM: TypeScript error in /opt/build/repo/src/App.tsx(43,25):
> 5:59:15 AM: Type '{ videoId: string | undefined; history: History<unknown>; location: Location<unknown>; match: match<{ videoId?: string | undefined; }>; staticContext?: StaticContext | undefined; }' is not assignable to type '{ videoId: string; }'.
> 5:59:15 AM:   Types of property 'videoId' are incompatible.
> 5:59:15 AM:     Type 'string | undefined' is not assignable to type 'string'.
> 5:59:15 AM:       Type 'undefined' is not assignable to type 'string'.  TS2322
> 5:59:15 AM:     41 |               render={props => {
> 5:59:15 AM:     42 |                 const { videoId } = props.match.params;
> 5:59:15 AM:   > 43 |                 return <routes.Videos {...props} {...{ videoId }} />;
> 5:59:15 AM:        |                         ^
> 5:59:15 AM:     44 |               }}
> 5:59:15 AM:     45 |             />
> 5:59:15 AM:     46 |             <Route
> 5:59:15 AM: npm ERR! code ELIFECYCLE
> 5:59:15 AM: npm ERR! errno 1
> 5:59:15 AM: npm ERR! mbt.guide@0.1.0 build: `CI= react-scripts build`
> 5:59:15 AM: npm ERR! Exit status 1
> 5:59:15 AM: npm ERR!
> 5:59:15 AM: npm ERR! Failed at the mbt.guide@0.1.0 build script.
> 5:59:15 AM: npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 5:59:15 AM: npm ERR! A complete log of this run can be found in:
> 5:59:15 AM: npm ERR!     /opt/buildhome/.npm/_logs/2021-03-03T00_29_15_282Z-debug.log

**Comment by JKHeadley** [comment]
@oormicreations git got mixed up somehow. I pushed the missing changes though some of them might be incorrect (they were all on the `videos` route). Please make sure all the styles/margins are correct on that route.

**Comment by JKHeadley** [comment]
Also pushed one more change that allows the `hide processed videos` button to show while the videos are loading since the loading can take a while. 😁

**Comment by oormicreations** [comment]
There are no errors, but somehow, for me, the video search is not working. It throws a red popup. I tried it via a proxy (kproxy.com) and same result. Its not working in the currently online app also. I do not know, perhaps it is ISP issue here?

> Also pushed one more change that allows the `hide processed videos` button to show while the videos are loading since the loading can take a while.

Yes, I guess we can turn it off by default now. There are not many unprocessed videos.

**Review by JKHeadley — APPROVED** [review]

**Comment by JKHeadley** [comment]
@oormicreations no worries, it looks like we just exceeded our daily youtube api quota. This is likely due to the way we fetch/search for unprocessed videos. I pushed an update to disable it by default. We will probably need to add a way to cache the list of unprocessed videos to help prevent this moving forward. I'll add an issue for it.

If everything else looks good feel free to go ahead and merge.

**Comment by oormicreations** [comment]
Ok it is done. As you said we will stick to small changes now. I will come back to coding in a few days.

**Pull request merged**
